### PR TITLE
[Ray Data] fix problem: to_pandas failed on datasets returned by from_spark

### LINKED ISSUE #32967: [Ray Data] to_pandas failed on datasets returned by from_spark
### What happened + What you expected to happen

such code
'''python
df = spark.range(100)
ds = ray.data.from_spark(df)
ds.to_pandas
'''
will fail because type check failed. Blocks saved by `from_spark` is bytes, although it's actually in arrow format. 

### Versions / Dependencies

ray 2.1.0

### Reproduction script

'''python
df = spark.range(100)
ds = ray.data.from_spark(df)
ds.to_pandas
'''

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by kira-lin**
## Related issue number
Closes #32967 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/arrow_block.py
Comment (clarkzinzow) [inline python/ray/data/_internal/arrow_block.py:129]: @kira-lin Why is this necessary? The existing [`TableBlockBuilder.add_block()` logic](https://github.com/ray-project/ray/blob/ff66e0ef1bc248a547a88f3eb96512d986b9ed57/python/ray/data/_internal/table_block.py#L77-L87) should convert the bytes into an Arrow table: (delegation)[https://github.com/ray-project/ray/blob/ff66e0ef1bc248a547a88f3eb96512d986b9ed57/python/ray/data/block.py#L394-L397], (conversion)[https://github.com/ray-project/ray/blob/ff66e0ef1bc248a547a88f3eb96512d986b9ed57/python/ray/data/_internal/arrow_block.py#L149-L152]

Is this an issue with the instance check in `TableBlockBuilder.add_block()`? If so, we can change the block type supplied by the `ArrowBlockBuilder` constructor to its parent's constructor to a `(pyarrow.Table, bytes)` tuple, since `isinstance()` checks work with a tuple of types, and `TableBlockBuilder._block_type` is only used for that instance check.

Comment (kira-lin) [reply python/ray/data/_internal/arrow_block.py]: oh, I see. I tried Union, didn't work. but this alone is not enough to solve the problem. Please see the update.

**Comment by kira-lin** [comment]
@amogkam RayDP save these blocks in java, and these blocks are in arrow stream format. But due to cross language serialization, the type information is lost, and can only be saved as bytes. 

I submitted a PR #20242 to add arrow back as a serialization format long time ago. Now serialization 2.0 is not on track, maybe we can just update and merge it? What do you think @jovany-wang @ericl

**Comment by jovany-wang** [comment]
Hi @kira-lin , I think it's okay to merge https://github.com/ray-project/ray/pull/20242 as the short term plan to unblock your issue.

And for the long term plan, I don't know whether the plugable serialization framework(serialization2.0) is indeed needed in community. @ericl is it a high priority item?

**Comment by jovany-wang** [comment]
@kira-lin Are you going to reopn https://github.com/ray-project/ray/pull/20242 or submit a new one? I can do the review for that.

**Comment by kira-lin** [comment]
@jovany-wang I'll probably submit a new one. I need to finish things I'm currently working on, so that'll be at least April

**Comment by jovany-wang** [comment]
@kira-lin Sounds good. Feel free to bother me if it's needed.

**Comment by kira-lin** [comment]
@clarkzinzow The failure in CI seems to be not related to my PR, is it?

**Comment by kira-lin** [comment]
can we merge this first? @clarkzinzow

File: python/ray/data/_internal/delegating_block_builder.py
Comment (clarkzinzow) [inline python/ray/data/_internal/delegating_block_builder.py:63]: Seems like we can unify these two paths instead of special-casing for `ArrowBlockAccessor`.
```suggestion
        self._builder.add_block(accessor.to_block())
```

**Review by clarkzinzow — COMMENTED** [review]
Looks good to merge after removing the special `ArrowBlockAccessor` case in the block builder!

**Comment by clarkzinzow** [comment]
@kira-lin Looks like lint is failing due to an unused import in `delegating_block_builder.py`: https://buildkite.com/ray-project/oss-ci-build-pr/builds/16356#0187209f-6d76-426c-9666-69d900445ec7

**Comment by kira-lin** [comment]
> @kira-lin Looks like lint is failing due to an unused import in delegating_block_builder.py: https://buildkite.com/ray-project/oss-ci-build-pr/builds/16356#0187209f-6d76-426c-9666-69d900445ec7

Addressed

**Review by clarkzinzow — APPROVED** [review]
Implementation and CI looks good!

**Comment by zhe-thoughts** [comment]
@clarkzinzow quick note, we are in feature freeze. Please tag me (and I will approve) before merging to `master`. Thanks!

**Pull request merged**
